File: 357-count-numbers-with-unique-digits/count-numbers-with-unique-digits.java
class Solution {
    public int countNumbersWithUniqueDigits(int n) {
        int num = (int)Math.pow(10, n)-1;
        return count(String.valueOf(num), 0, 1, 1, 0, 0);
    }
    private int count(String s, int i, int t, int lz, int r, int mask) {
        if(i == s.length()) {
            if(r == 0) {
                return 1;
            }
            return 0;
        }
        int lb = 0;
        int ub = t==1?(s.charAt(i)-'0'):9;
        int ans = 0;
        for(int digit = lb; digit <= ub; digit++) {
            int newTight = (t==1 && digit==ub)?1:0;
            int newLz = (lz==1 && digit==0)?1:0;
            int newR = r;
            if((mask & (1<<digit)) != 0) {
                newR = 1;
            }
            int newMask = mask;
            if(digit == 0 && lz == 0) {
                newMask = mask | (1<<digit);
            }
            else if(digit > 0) {
                newMask = mask | (1<<digit);
            }
            ans += count(s, i+1, newTight, newLz, newR, newMask);
        }
        return ans;
    }
}

File: 357-count-numbers-with-unique-digits/count-numbers-with-unique-digits.cpp
class Solution {
public:
    int countNumbersWithUniqueDigits(int n) {

        int x = pow(10, n) - 1;
        string s = to_string(x);

        return count(s, 0, 1, 0);
    }
    int count(string& s, int idx, int t, int mask) {

        if (idx == s.length()) {
            return 1;
        }

        int lb = 0;
        int ub = t == 1 ? (s[idx] - '0') : 9;
        int res = 0;

        for (int d = lb; d <= ub; d++) {

            if ((mask & (1 << d)) != 0) {
                continue;
            }

            int newMask = 0;
            int newT = (t == 1 && d == ub) ? 1 : 0;

            if (mask != 0 || d != 0) {
                newMask = (mask | (1 << d));
            }

            res += count(s, idx + 1, newT, newMask);
        }

        return res;
    }
};